data/src/main/java/com/microsoft/azure/kusto/data/instrumentation/MonitoredActivity.java

```diff
@@ -0,0 +1,67 @@
+package com.microsoft.azure.kusto.data.instrumentation;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class MonitoredActivity {
+    public static void invoke(Runnable runnable, String nameOfSpan) {
+        invoke(runnable, nameOfSpan, new HashMap<>());
+    }
+
+    public static void invoke(Runnable runnable, String nameOfSpan, Map<String, String> attributes) {
+        try (Tracer.Span span = Tracer.startSpan(nameOfSpan, attributes)) {
+            runnable.run();
+        }
+    }
+
+    public static <T, U extends Exception> T invoke(SupplierOneException<T, U> supplier, String nameOfSpan) throws U {
+        return invoke((SupplierTwoExceptions<T, U, U>) supplier::get, nameOfSpan, new HashMap<>());
+    }
+
+    public static <T, U extends Exception> T invoke(SupplierOneException<T, U> supplier, String nameOfSpan, Map<String, String> attributes) throws U {
+        return invoke((SupplierTwoExceptions<T, U, U>) supplier::get, nameOfSpan, attributes);
+    }
+
+    public static <T, U1 extends Exception, U2 extends Exception> T invoke(SupplierTwoExceptions<T, U1, U2> supplier, String nameOfSpan) throws U1, U2 {
+        return invoke(supplier, nameOfSpan, new HashMap<>());
+    }
+
+    public static <T, U1 extends Exception, U2 extends Exception> T invoke(SupplierTwoExceptions<T, U1, U2> supplier, String nameOfSpan,
+            Map<String, String> attributes) throws U1, U2 {
+        try (Tracer.Span span = Tracer.startSpan(nameOfSpan, attributes)) {
+            try {
+                return supplier.get();
+            } catch (Exception e) {
+                span.addException(e);
+                throw e;
+            }
+        }
+    }
+
+    public static <T, U extends Exception> T invoke(FunctionOneException<T, Tracer.Span, U> function, String nameOfSpan) throws U {
```

Comment:
Document maybe that this is the variant that lets the inner mehtod use the span object